src/rust/src/orchard_ffi.rs

```diff
@@ -65,3 +73,108 @@ pub extern "C" fn orchard_bundle_serialize(
         }
     }
 }
+
+/// A signature within an authorized Orchard bundle.
+#[derive(Debug)]
+struct BundleSignature {
+    /// The signature item for validation.
+    signature: redpallas::batch::Item<SpendAuth, Binding>,
+}
+
+/// Batch validation context for Orchard.
+pub struct BatchValidator {
+    signatures: Vec<BundleSignature>,
+}
+
+impl BatchValidator {
+    fn new() -> Self {
+        BatchValidator { signatures: vec![] }
+    }
+}
+
+/// Creates a RedPallas batch validation context.
+///
+/// Please free this when you're done.
+#[no_mangle]
+pub extern "C" fn orchard_batch_validation_init() -> *mut BatchValidator {
+    let ctx = Box::new(BatchValidator::new());
+    Box::into_raw(ctx)
+}
+
+/// Frees a RedPallas batch validation context returned from
+/// [`orchard_batch_validation_init`].
+#[no_mangle]
+pub extern "C" fn orchard_batch_validation_free(ctx: *mut BatchValidator) {
+    if !ctx.is_null() {
+        drop(unsafe { Box::from_raw(ctx) });
+    }
+}
+
+/// Adds an Orchard bundle to this batch.
+///
+/// Currently, only RedPallas signatures are batch-validated.
+#[no_mangle]
+pub extern "C" fn orchard_batch_add_bundle(
+    batch: *mut BatchValidator,
+    bundle: *const Bundle<Authorized, Amount>,
+    txid: *const [u8; 32],
+) {
+    let batch = unsafe { batch.as_mut() };
+    let bundle = unsafe { bundle.as_ref() };
+    let txid = unsafe { txid.as_ref() }.cloned().map(TxId::from_bytes);
+
+    match (batch, bundle, txid) {
+        (Some(batch), Some(bundle), Some(txid)) => {
+            for action in bundle.actions().iter() {
+                batch.signatures.push(BundleSignature {
+                    signature: action
+                        .rk()
+                        .create_batch_item(action.authorization().clone(), txid.as_ref()),
+                });
+            }
+
+            batch.signatures.push(BundleSignature {
+                signature: bundle.binding_validating_key().create_batch_item(
+                    bundle.authorization().binding_signature().clone(),
+                    txid.as_ref(),
+                ),
+            });
+        }
+        (_, _, None) => error!("orchard_batch_add_bundle() called without txid!"),
+        (Some(_), None, Some(txid)) => debug!("Tx {} has no Orchard component", txid),
+        (None, Some(_), _) => debug!("Orchard BatchValidator not provided, assuming disabled."),
+        (None, None, _) => (), // Boring, don't bother logging.
```

Comment:
This doesn't seem boring to me, it should never happen right?

Reply:
This case means that a disable batch validator was given a transaction with no Orchard component. I can see that happening regularly.

Reply:
(specifically, during the earlier validation passes in `ProcessNewBlock`, since we only want to check signatures once in the last pass)